### PR TITLE
React: Set `react-docgen` to default TS docgen

### PR DESCRIPTION
**Pull request by shilman**
Telescopes on #23825

## What I did

Set `react-docgen` as the default so we can look at Chromatic diffs easily. Won't merge until:
- [x]  We are comfortable with the limitations, e.g. [lack of support for `React.FC`](https://github.com/reactjs/react-docgen/issues/387) is probably a deal-breaker
- [x] 8.0 dev starts

Also needs:
- [x] Automigration

### Testing

- [ ] See attached Chromatic diffs
- [ ] Will add more test cases TBD

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24165-sha-acaa9f93`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24165-sha-acaa9f93). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24165-sha-acaa9f93`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24165-sha-acaa9f93) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/default-to-react-docgen`](https://github.com/storybookjs/storybook/tree/shilman/default-to-react-docgen) |
| **Commit** | [`acaa9f93`](https://github.com/storybookjs/storybook/commit/acaa9f93dea09d179a9426e87b8d9b7ca072218f) |
| **Datetime** | Mon Oct 23 07:43:10 UTC 2023 (`1698046990`) |
| **Workflow run** | [6610198728](https://github.com/storybookjs/storybook/actions/runs/6610198728) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24165`_
</details>

### CODE REVIEW

**Comment by shilman** [comment]
@valentinpalkovic will evaluate [npmjs.com/package/typescript-react-function-component-props-handler](https://www.npmjs.com/package/typescript-react-function-component-props-handler). If it works well, it could be a good candidate to incorporate into the default settings. cc @danez

File: code/renderers/react/template/stories/ts-argtypes.stories.tsx [hunk 1]
```diff
@@ -78,14 +78,14 @@ export const TsComponentProps = { parameters: { component: TsComponentPropsCompo
...
-  }
-};
+// const addChromaticIgnore = async (element: HTMLElement) => {
+//   const row = element.parentElement?.parentElement;
+//   if (row?.nodeName === 'TR') {
+//     row.setAttribute('data-chromatic', 'ignore');
+//   } else {
+//     throw new Error('the DOM structure changed, please update this test');
+//   }
+// };
```
Comment (ndelangen) [inline code/renderers/react/template/stories/ts-argtypes.stories.tsx:88]: ?

Comment (shilman) [reply code/renderers/react/template/stories/ts-argtypes.stories.tsx:88]: won't merge with this. this is code that makes ignore regions for all the key data that we want to test. will probably delete before the merge & then re-add it if we see test flake

**Comment by danez** [comment]
I released version 7.0 of `react-docgen` which supports `React.FC` and all the other helpers like `React.PropsWithChildren`.

**Comment by shilman** [comment]
@danez Thanks so much for taking care of this!! ❤️  I've tested it out as part of https://github.com/storybookjs/storybook/pull/24530 and it's working great -- we'll release the upgrade as part of SB7.6.

As for making `react-docgen` the default, I'll add some code to help users migrate gracefully from `react-docgen-typescript` before merging. I'll let you know if we run into any serious bugs as we alpha test this with users, and will generally keep you posted on the progress. This change is going to make a lot of people really happy!!!

File: .nvmrc [hunk 2]
```diff
@@ -1 +1 @@
-18.18.2
+16.20
```
Comment (ndelangen) [inline .nvmrc:1]:
```suggestion
18.18.2
```

File: code/addons/docs/react/README.md [hunk 3]
```diff
@@ -108,7 +108,7 @@ To do so for all stories, update `.storybook/preview.js`:
 export const parameters = { docs: { story: { inline: false } } };
 ```
 
-## TypeScript props with `react-docgen`
+## TypeScript props with `react-docgen-typescript`
 
 If you're using TypeScript, there are two different options for generating props: `react-docgen-typescript` (default) or `react-docgen`.
```
Comment (IanVS) [inline code/addons/docs/react/README.md:113]: This needs to be updated too, looks like.

File: code/lib/core-server/src/presets/common-preset.ts [hunk 4]
```diff
@@ -134,7 +134,7 @@ export const previewBody = async (base: any, { configDir, presets }: Options) =>
 export const typescript = () => ({
   check: false,
   // 'react-docgen' faster but produces lower quality typescript results
```
[the remark this answers is not in the document]
Comment (shilman) [reply code/lib/core-server/src/presets/common-preset.ts]: good catch, thanks!

**Comment by csantos1113** [comment]
I'm curious what is the foundational difference between `react-docgen` and `react-docgen-typescript` and when I should use one or another?

**Comment by ndelangen** [comment]
@shilman can this PR proceed?

**Comment by shilman** [comment]
Not yet. There are still some chromatic diffs that I'd like to investigate. @ndelangen

**Comment by csantos1113** [comment]
> @csantos1113 `react-docgen` is a lot faster, because it doesn't use `tsc`-APIs.
> 
> But those `tsc`-APIs provide output akin to the actual types, whereas `react-docgen`'s output can be less correct.

@ndelangen - Thanks for the reply; I changed it to `react-docgen` and while it's significantly faster, I'm losing lot of useful context from the types and inline docs


left is `react-docgen-typescript`, right is `react-docgen`

Screenshot 2023-11-29 at 11 45 49 AM

Screenshot 2023-11-29 at 11 46 30 AM

One of these props in code:
image

----

Sorry for hijacking this PR with this question, _I could also open a issue about this topic_, but I think it's sort of relevant for this PR because the default behavior will change in SB 8.0, if we make `react-docgen` the default

**Comment by valentinpalkovic** [comment]
@csantos1113 Can you tell us with which Storybook version you have tried this and whether you are using a webpack5 or vite based project?

**Comment by csantos1113** [comment]
> @csantos1113 Can you tell us with which Storybook version you have tried this and whether you are using a webpack5 or vite based project?

Certainly @valentinpalkovic !

- storybook packages: `7.6.1`
- using `@storybook/react-vite`
- my codebase is written in typescript

Here a simplified code example:
image

#### `types.ts`
```ts
export enum ButtonType {
  Button = 'button',
  Reset = 'reset',
  Submit = 'submit'
}

export type ButtonTestProps = {
  /**
   * Optional button type for the component
   */
  type?: ButtonType;
};
```

#### `index.tsx`

```tsx
import { ButtonTestProps } from './types';

export const ButtonTest = (props: ButtonTestProps) => {
  return <button {...props}>Test</button>;
};
```

#### `index.stories.ts`
```ts
import { Meta, StoryObj } from '@storybook/react';

import { ButtonTest } from '.';
import { ButtonType } from './types';

const meta = {
  args: {
    type: ButtonType.Button
  },
  component: ButtonTest
} satisfies Meta<typeof ButtonTest>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};
```

### ✅ 😍 Results with `react-docgen-typescript`

image

### 🤕 Results with `react-docgen` 

image

Notice the "Description" and "Control"

### 😞 And it's even worst if I use path aliases (for example `~/button-test/` instead of `./`) on the imports

- The path alias is configured in my `tsconfig.json`( `~/*` => `./src/*`)
- I have https://github.com/aleclarson/vite-tsconfig-paths in `vite.config.ts` for alias to resolve in vite
- path aliases are pretty common

image

Notice the "Description" is gone

### Diff between `react-docgen` and `react-docgen-typescript`

I understand there are foundational differences between the two libraries, and `react-docgen` has a bunch of open TypeScript related issues: https://github.com/reactjs/react-docgen/issues?q=is%3Aissue+is%3Aopen+typescript

---

It is understandable that making these two libraries work "the same" is not responsibility of the Storybook team; _but_ if Storybook changes the default to `react-docgen`, a lot of typescript users like myself will be impacted; so perhaps a warning/notification should be added somewhere in the migration guides at a minimum.

Note: I know `8.0` is in is early stages, just adding my early feedback too 😅

File: code/lib/cli/src/automigrate/fixes/react-docgen.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    return dedent`
+      You have "typescript.reactDocgenTypescriptOptions" configured in your main.js,
+      but "typescript.reactDocgen" is unset.
+      
+      In Storybook 8.0, we changed the default React docgen analysis from 
+      "react-docgen-typescript" to "react-docgen", which dramatically faster
+      but doesn't handle all TypeScript constructs.
+
+      We can update your config to continue to use "react-docgen-typescript",
+      though we recommend giving "react-docgen" or a much faster dev experience.
```
Comment (shilman) [inline code/lib/cli/src/automigrate/fixes/react-docgen.ts:33]:
```suggestion
      though we recommend giving "react-docgen" for a much faster dev experience.
```

**Pull request merged**